File: 15.py
om = [[int(x)  for x in line[:-1]] for line in open("15.txt").readlines()]
m = []
for li in range(5):
    for lj in range(5):
        o = li + lj
        f = lambda x: x%9 if x%9 != 0 else 9
        for i in range(len(om)):
            if lj == 0:
                m.append([])
            for j in range(len(om[i])):
                m[li * len(om) + i].append(f(om[i][j] + o))

q = [(0, 0, 0, 0, 0)]
v = [[False for j in range(len(m[i]))] for i in range(len(m))]
d = [[-1 for j in range(len(m[i]))] for i in range(len(m))]
p = [[(i, j) for j in range(len(m[i]))] for i in range(len(m))]
while len(q):
    q.sort(key=lambda x: x[0])
    dst, i, j, pi, pj = q.pop(0)
    if d[i][j] >= 0:
        continue
    d[i][j] = dst
    p[i][j] = (pi, pj)
    v[i][j] = True
    if (i == len(m)-1 and j == len(m[i])-1):
        print(dst)
        path = [(i, j)]
        while not(pi == p[pi][pj][0] and pj == p[pi][pj][1]):
            path.append((pi, pj))
            pi, pj = p[pi][pj]
        path.append((0, 0))
        open("15path.txt", "w").write("\n".join(["".join('#' if (x, y) in path else ' ' for x in range(j+1)) for y in range(i+1)]))
        # break
    for ni, nj in [(-1, 0), (0, -1), (0, 1), (1, 0)]:
        if (i + ni) >= 0 and i + ni < len(m) and (j + nj) >= 0 and j + nj < len(m[i + ni]) and not v[i+ni][j+nj]:
            q.append((dst+m[i+ni][j+nj], i + ni, j+nj, i, j))

open("15d.txt", "w").write("\n".join(["|".join(f'{x:>2}' for x in r) for r in d]))

File: 2022/15.jl
function merge_intervals!(inters::Vector{Pair{Int64, Int64}})
    back_ix = length(inters)
    while back_ix > 1
        left, right = popat!(inters, back_ix)
        merged = false
        for i in range(1, length(inters))
            fleft, fright = inters[i]
            if left - fright > 1 || fleft - right > 1
                # not touching
                continue
            else
                inters[i] = Pair(min(fleft, left), max(fright, right))
                merged = true
                break
            end
        end
        if !merged
            push!(inters, Pair(left, right))
        end
        back_ix -= 1
    end
    sort!(inters)
end

function main()
    lines = readlines("15.txt")

    y = 2e6
    y = 10
    bound = 4000000
    bound = 20
    n_beacons_on_y = 0
    n = length(lines)
    y_inters = Vector{Pair{Int64, Int64}}()
    top_left_ys  = Array{Union{Nothing, Int64}, 1}(nothing, n)
    top_right_ys = Array{Union{Nothing, Int64}, 1}(nothing, n)
    bot_left_ys  = Array{Union{Nothing, Int64}, 1}(nothing, n)
    bot_right_ys = Array{Union{Nothing, Int64}, 1}(nothing, n)

    up_pairs = Vector{Int64}()
    down_pairs = Vector{Int64}()

    for (i, line) in enumerate(lines)
        sx, sy, bx, by = [parse(Int64, line[coord]) for coord in findall(r"-?\d+", line)]
        if by == y
            n_beacons_on_y += 1
        end
        # part 1
        man_dist = abs(sx - bx) + abs(sy - by)
        range = max(0, man_dist - abs(sy - y))
        # println(man_dist, " ", vert_dist)
        push!(y_inters, Pair(sx - range, sx + range))
        merge_intervals!(y_inters)

        # part 2
        # y intercept of boundaries that are outside of man_dist
        top_left_ys[i]  = sy + (man_dist + 1) - sx
        top_right_ys[i] = sy + (man_dist + 1) + sx
        bot_left_ys[i]  = sy - (man_dist + 1) + sx
        bot_right_ys[i] = sy - (man_dist + 1) - sx

        if top_left_ys[i] in bot_right_ys[i]
            push!(up_pairs, top_left_ys[i])
        end
        if top_right_ys[i] in bot_left_ys[i]
            push!(down_pairs, top_right_ys[i])
        end
        if bot_left_ys[i] in top_right_ys[i]
            push!(down_pairs, bot_left_ys[i])
        end
        if bot_right_ys[i] in top_left_ys[i]
            push!(up_pairs, bot_right_ys[i])
        end
    end
    println(y_inters)
    left, right = y_inters[1]
    println(right - left + 1 - n_beacons_on_y)

    println(up_pairs)
    println(down_pairs)
    println(top_left_ys)
    println(top_right_ys)
    println(bot_left_ys)
    println(bot_right_ys)
    println(intersect(top_left_ys, top_right_ys, bot_left_ys, bot_right_ys))
end

@time main()